### PR TITLE
expression: fix wrong result type for greatest/least

### LINKED ISSUE #29019: Query with greatest/least function returns inconsistent result with mysql
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

https://asktug.com/t/topic/183612

```
create table test.test_GREATEST(
    id bigint(20),
    datetime_1 datetime,
    datetime_2 datetime,
    datetime3_1 datetime(3),
    datetime3_2 datetime(3)
);

insert into test.test_GREATEST values(1,'2021-10-20 10:10:10.000','2021-10-20 10:10:10.000','2021-10-20 10:10:10.000','2021-10-20 10:10:10.000');

select
    *
from(
    select
        id,
        GREATEST(tg.datetime_1,tg.datetime_2) datetimes
    from test.test_GREATEST tg
) t
where t.datetimes = '2021-10-20 10:10:10.000';

```

### 2. What did you expect to see? (Required)
+------+---------------------+
| id   | datetimes           |
+------+---------------------+
|    1 | 2021-10-20 10:10:10 |
+------+---------------------+
1 row in set (0.00 sec)

### 3. What did you see instead (Required)
Empty set (0.00 sec)

### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### What problem does this PR solve?

Issue Number: close #29019

Problem Summary: 
1. When all arguments types for greatest/least are datetime, we deduce the result type as string incorrectly.

What's changed:
1. So we need to add four new function sigs for duration and time(builtinGreatestDurationSig, builtinGreatestTimeSig, builtinLeastDurationSig, builtinLeastTimeSig).
2. Also we change the original sig to builtinGreatestCmpStringAsDatetime/builtinLeastCmpStringAsDatetime, because the original sig name is not reasonable.

TODO:
1. When arguments of greatest/least are mixed with time, duration, string, the result is not compitable with mysql. Such as `select least(cast('-800:10:10' as time), cast('2021-10-10 10:10:10.1234' as datetime), '99999999999') from t1;`
### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
fix wrong result type for greatest/least
```

### CODE REVIEW

File: expression/builtin_compare.go [hunk 1]
```diff
@@ -706,8 +754,17 @@ func (c *leastFunctionClass) getFunction(ctx sessionctx.Context, args []Expressi
 		sig = &builtinLeastDecimalSig{bf}
 		sig.setPbCode(tipb.ScalarFuncSig_LeastDecimal)
 	case types.ETString:
-		sig = &builtinLeastStringSig{bf}
-		sig.setPbCode(tipb.ScalarFuncSig_LeastString)
+		if cmpStringAsDatetime {
+			sig = &builtinLeastCmpStringAsDatetimeSig{bf}
+			// TODO
+			// sig.setPbCode(tipb.ScalarFuncSig_GreatestString)
+		} else {
```
Comment (guo-shaoge) [inline expression/builtin_compare.go:761]: Wait on this tipb [pr](https://github.com/pingcap/tipb/pull/246)

File: expression/builtin_compare.go [hunk 2]
```diff
@@ -487,8 +486,17 @@ func (c *greatestFunctionClass) getFunction(ctx sessionctx.Context, args []Expre
 		sig = &builtinGreatestDecimalSig{bf}
 		sig.setPbCode(tipb.ScalarFuncSig_GreatestDecimal)
 	case types.ETString:
-		sig = &builtinGreatestStringSig{bf}
-		sig.setPbCode(tipb.ScalarFuncSig_GreatestString)
+		if cmpStringAsDatetime {
+			sig = &builtinGreatestCmpStringAsDatetimeSig{bf}
+			// TODO
+			// sig.setPbCode(tipb.ScalarFuncSig_GreatestString)
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go]: uncomment this?

**Review by XuHuaiyu — APPROVED** [review]
LGTM

File: expression/builtin_compare_vec.go [hunk 3]
```diff
@@ -815,3 +815,131 @@ func (b *builtinGreatestStringSig) vecEvalString(input *chunk.Chunk, result *chu
...
+		if err := b.args[argIdx].VecEvalTime(b.ctx, input, buf); err != nil {
+			return err
+		}
+		result.MergeNulls(buf)
+		resTimes := result.Times()
+		for rowIdx := 0; rowIdx < n; rowIdx++ {
+			if result.IsNull(rowIdx) {
+				continue
+			}
+			argTime := buf.GetTime(rowIdx)
```
[the remark this answers is not in the document]
Comment (guo-shaoge) [reply expression/builtin_compare_vec.go]: Done

File: expression/builtin_compare_vec.go [hunk 4]
```diff
@@ -815,3 +815,131 @@ func (b *builtinGreatestStringSig) vecEvalString(input *chunk.Chunk, result *chu
...
```
[the remark this answers is not in the document]
Comment (guo-shaoge) [reply expression/builtin_compare_vec.go]: Done

File: expression/builtin_compare.go (hunk 2 above)
Comment (guo-shaoge) [reply expression/builtin_compare.go]: Done

**Review by wshwsh12 — APPROVED** [review]

**Comment by guo-shaoge** [comment]
Also need to cherry pick #31037, otherwise may introduce new problem #29019 .

**Pull request merged**